0.15.2 auto-inject applies the preview plugin to KMP modules with no previews, failing the render on a variant ambiguity (:meshcore-mobile:androidRuntimeClasspath)

## Summary

As of **0.15.2**, the CLI's auto-inject applies `ee.schimke.composeai.preview` to **every** `com.android.kotlin.multiplatform.library` module — including modules that declare **no `@Preview` functions**. In a multi-module KMP project this pulls non-UI library modules into preview discovery, and resolving one of them (`:meshcore-mobile`, which has no previews) fails the whole compose render on a **variant ambiguity**, which the `apply` action then reports as an empty `_previews.json` (red CI).

This is the flip side of the 0.15.2 change that (correctly) lets KMP-android-library modules be auto-injected without an explicit `plugins {}` entry — it's just too broad: it injects into library modules that have nothing to render.

## Repro

A project (meshcore-mobile) with several `com.android.kotlin.multiplatform.library` modules, only one of which (`:meshcore-components`) has `@Preview`s. With `apply@v0.15.2` / CLI 0.15.2:

```
##[group]compose pipeline
Found preview modules: app, meshcore-components, meshcore-core, meshcore-data, meshcore-mobile, meshcore-transport-ble, meshcore-transport-usb, wear
FAILURE: Build failed with an exception.
* What went wrong:
> Could not resolve all dependencies for configuration ':meshcore-mobile:androidRuntimeClasspath'.
> Ambiguity errors are explained in more detail at …#sub:variant-ambiguity.
Render failed
compose pipeline: no _previews.json — workspace has no compose modules, skipping.
```

`:meshcore-core`, `:meshcore-data`, `:meshcore-mobile`, `:meshcore-transport-*` have **no previews** — before 0.15.2 (or with the plugin applied explicitly to just `:meshcore-components`) the discovered set was `app, wear, meshcore-components` and the compose render was green.

## Notes for diagnosis

- `./gradlew :meshcore-mobile:dependencies --configuration androidRuntimeClasspath` **resolves fine** (no conflict). So the failure is in the **render classpath's variant/artifact selection** under the attributes the plugin requests, not a version conflict.
- The CLI **swallows the detailed Gradle ambiguity message** — only the one-line `Could not resolve … Ambiguity errors are explained …` summary is printed; the `Cannot choose between the following variants … because …` block (the part that names the ambiguous dependency + candidate variants) never reaches stdout. Surfacing that block (or a `--stacktrace`/verbose passthrough) would make this kind of failure self-diagnosable.

## Impact

The auto-inject improvement makes 0.15.2 strictly worse for this repo than 0.15.1-with-explicit-plugin: the existing `compose-preview` baseline workflow (which was green) now fails the whole compose pipeline because a non-preview module's render classpath can't be resolved. There's no per-module opt-out for the compose pipeline (only `a11y-*-modules` / `notifications-*-modules`), so a consumer can't scope it back to the modules that actually have previews.

## Suggestions (any one would unblock)

1. **Only inject / discover modules that actually declare `@Preview` functions** — a library module with no previews shouldn't register the render tasks at all.
2. **Fail-soft per module**: if one module's render classpath can't resolve, skip that module (warn) instead of failing the whole compose pipeline / emitting an empty `_previews.json`.
3. **Expose a compose-pipeline module allow/deny** (like `a11y-module` / `notifications-skip-modules`) so consumers can scope discovery to `app,wear,meshcore-components`.
4. **Surface the full Gradle variant-ambiguity message** from the CLI so the underlying conflict is diagnosable.

Context: came up adopting design-parity on the CMP desktop backend in meshcore-mobile (`:meshcore-components` is a desktop preview module). Happy to provide a full `--scan` / stacktrace of the `:meshcore-mobile:androidRuntimeClasspath` resolution if useful. Sibling to #1843 / #1844 / #1845 / #1846 from the same adoption.

--- a/gradle-plugin/src/main/kotlin/ee/schimke/composeai/plugin/ComposePreviewTasks.kt
+++ b/gradle-plugin/src/main/kotlin/ee/schimke/composeai/plugin/ComposePreviewTasks.kt
@@ -89,6 +89,28 @@ internal object ComposePreviewTasks {
     return rendererConfig
   }
 
+  /**
+   * Whether a desktop-routed module whose resolved runtime config is [configName] can actually be
+   * rendered by the JVM desktop renderer.
+   *
+   * The renderer is a JVM process, so it needs JVM-flavoured Compose on the classpath. A pure
+   * `com.android.kotlin.multiplatform.library` module with no `jvm("desktop")` target exposes only
+   * `androidRuntimeClasspath` — carrying `*-android` Compose AARs that reference
+   * `android.os.Parcelable` — and forcing any artifact view on its single AGP `android` variant
+   * trips a variant ambiguity that, left to propagate, hard-fails the whole `composePreviewRender`
+   * pipeline for every module (issue #1852: a no-preview library module like `:meshcore-mobile`
+   * sinks the green modules' render too).
+   *
+   * Such a module is treated as non-renderable: the desktop tasks skip it (fail-soft per module,
+   * with a one-line warning) instead of resolving its classpath and failing. Adding a
+   * `jvm("desktop")` target — the canonical cmp-shared layout — surfaces `desktopRuntimeClasspath`
+   * and flips this back to `true`.
+   *
+   * `internal` + pure so it can be unit-tested without standing up AGP/KGP.
+   */
+  internal fun isDesktopRenderableConfig(configName: String): Boolean =
+    configName != "androidRuntimeClasspath"
+
   fun registerDesktopTasks(project: Project, extension: PreviewExtension) {
     val previewOutputDir = project.layout.buildDirectory.dir("compose-previews")
 
@@ -164,7 +186,15 @@ internal object ComposePreviewTasks {
         // aborts on such a module. JVM/desktop configs stay strict.
         lenientWhenAndroidOnlyFallback = true,
       ) {
-        onlyIf { extension.enabled.get() }
+        // Skip discovery for a non-renderable pure-KMP-Android module (only
+        // androidRuntimeClasspath,
+        // no jvm("desktop") target) so it reports nothing instead of being pulled into the pipeline
+        // — keeps it consistent with the render task, which skips the same module (issue #1852).
+        // Computed at task-realization time, when the kotlin{} block has configured and
+        // desktopRuntimeClasspath (if any) exists; captured as a Boolean so onlyIf stays
+        // config-cache-safe.
+        val renderable = isDesktopRenderableConfig(resolveDependencyConfigName())
+        onlyIf { extension.enabled.get() && renderable }
         // `compileAndroidMain` is the lifecycle task for the KMP-Android target's `main`
         // compilation (which depends on the underlying `compileAndroidMainKotlin`-shaped Kotlin
         // compile under the hood). Use lazy matching so compile tasks registered after this plugin
@@ -192,7 +222,22 @@ internal object ComposePreviewTasks {
       )
     val renderTask =
       project.tasks.register("composePreviewRender", RenderPreviewsTask::class.java) {
-        onlyIf { extension.enabled.get() }
+        // A pure KMP-Android module with no jvm("desktop") target can't be rendered by the JVM
+        // desktop renderer; resolving its androidRuntimeClasspath would trip a variant ambiguity
+        // and hard-fail the whole pipeline (issue #1852). Skip it (fail-soft per module) and warn
+        // once — realization time is after the kotlin{} block configures, so the config name is
+        // settled here; capture as a Boolean to keep onlyIf config-cache-safe.
+        val renderable = isDesktopRenderableConfig(resolveDependencyConfigName())
+        if (!renderable) {
+          logger.warn(
+            "compose-preview: skipping module '${project.path}' — it applies " +
+              "com.android.kotlin.multiplatform.library but has no jvm(\"desktop\") target, so the " +
+              "Compose Multiplatform Desktop renderer can't render its *-android Compose artifacts. " +
+              "Add a jvm(\"desktop\") target to render its androidMain previews (see " +
+              "compose-preview/references/cmp-shared.md)."
+          )
+        }
+        onlyIf { extension.enabled.get() && renderable }
         previewsJson.set(previewOutputDir.map { it.file("previews.json") })
         outputDir.set(previewOutputDir.map { it.dir("renders") })
         // `@ScrollingPreview(modes = [LONG, GIF])` outputs land here (sibling of `renders/`).
@@ -211,8 +256,13 @@ internal object ComposePreviewTasks {
         // `Configuration` into the task's config-cache `__classpath__` field (the serialization
         // failure issue #1796 hit on the desktop validate guards). The empty view yields the same
         // default resolution — module + file dependencies alike — just lazily and serializably.
-        project.configurations.findByName(resolveDependencyConfigName())?.let {
-          renderClasspath.from(it.incoming.artifactView {}.files)
+        // Only wired when renderable: a non-renderable module's androidRuntimeClasspath would
+        // resolve to a variant ambiguity during input snapshotting even though onlyIf skips the
+        // action, so leave it off the classpath entirely (issue #1852).
+        if (renderable) {
+          project.configurations.findByName(resolveDependencyConfigName())?.let {
+            renderClasspath.from(it.incoming.artifactView {}.files)
+          }
         }
         renderClasspath.from(rendererConfig.incoming.artifactView {}.files)
         group = "compose preview"
@@ -411,8 +461,15 @@ internal object ComposePreviewTasks {
         candidates.firstOrNull { it != null && it.asFile.isDirectory }
       }
 
+    // A non-renderable pure-KMP-Android module has nothing the desktop renderer can pack; skip its
+    // bundle for consistency with the render/daemon tasks (issue #1852).
+    // `isDesktopRenderableConfig`
+    // only trips on the literal `androidRuntimeClasspath` fallback, so this is a no-op on the
+    // Android
+    // bundle path (which routes through `${variant}RuntimeClasspath`) and on real desktop modules.
+    val bundleRenderable = isDesktopRenderableConfig(depConfigName)
     project.tasks.register("composePreviewBundle", BundlePreviewTask::class.java) {
-      onlyIf { extension.enabled.get() }
+      onlyIf { extension.enabled.get() && bundleRenderable }
       previewsJson.set(previewOutputDir.map { it.file("previews.json") })
       moduleClassDirs.from(sourceClassDirs)
       moduleResourcesDir.set(moduleResourcesDirProvider)
@@ -559,6 +616,12 @@ internal object ComposePreviewTasks {
       "composePreviewDaemonStart",
       ee.schimke.composeai.plugin.daemon.DaemonBootstrapTask::class.java,
     ) {
+      // A non-renderable pure-KMP-Android module (only androidRuntimeClasspath, no jvm("desktop"))
+      // can't be driven by the desktop daemon either, and resolving its runtime classpath would
+      // trip the same variant ambiguity — skip the descriptor emission and leave its config off
+      // the daemon `-cp` (issue #1852).
+      val renderable = isDesktopRenderableConfig(dependencyConfigName())
+      onlyIf { renderable }
       modulePath.set(project.path)
       // Desktop daemons have no AGP variant. The string is surfaced in `daemon-launch.json`'s
       // `variant` field for debug/log purposes only — VS Code's `daemonProcess.ts` doesn't key
@@ -578,7 +641,8 @@ internal object ComposePreviewTasks {
         project = project,
         extension = extension,
         task = this,
-        userRuntimeConfig = project.configurations.findByName(dependencyConfigName()),
+        userRuntimeConfig =
+          if (renderable) project.configurations.findByName(dependencyConfigName()) else null,
       )
       // `:daemon:desktop`'s `DaemonMain` and `:daemon:android`'s `DaemonMain` share the FQN
       // intentionally (see the kdoc on `daemon/desktop/.../DaemonMain.kt`). The desktop classes
@@ -601,8 +665,12 @@ internal object ComposePreviewTasks {
       classpath.from(sourceResourceDirs)
       // User's runtime classpath (Compose Multiplatform deps + transitive Kotlin libraries).
       // Lazy artifact view (not the raw `Configuration`) for config-cache serializability — #1796.
-      project.configurations.findByName(dependencyConfigName())?.let {
-        classpath.from(it.incoming.artifactView {}.files)
+      // Skipped for a non-renderable module so its ambiguous androidRuntimeClasspath is never
+      // resolved (issue #1852).
+      if (renderable) {
+        project.configurations.findByName(dependencyConfigName())?.let {
+          classpath.from(it.incoming.artifactView {}.files)
+        }
       }
 
       // Desktop daemons don't run inside Robolectric, so the AGP-side `--add-opens` flags don't
@@ -771,6 +839,12 @@ internal object ComposePreviewTasks {
     toolClasspath: org.gradle.api.artifacts.Configuration,
   ): TaskProvider<ValidateComposePreviewClasspathTask> =
     project.tasks.register(taskName, ValidateComposePreviewClasspathTask::class.java) {
+      // The guard runs as a dependsOn of the render/daemon task, so it would resolve
+      // androidRuntimeClasspath (and fail on its variant ambiguity) BEFORE the render task's own
+      // onlyIf can skip a non-renderable module. Gate it the same way: skip the guard and don't
+      // wire the consumer config when the module isn't desktop-renderable (issue #1852).
+      val renderable = isDesktopRenderableConfig(dependencyConfigName())
+      onlyIf { renderable }
       platform.set("desktop")
       // Feed the @Classpath FileCollection a lazily-resolved, content-keyed FileCollection
       // (`incoming.artifactView { }.files`) instead of the raw `Configuration`.
@@ -784,8 +858,10 @@ internal object ComposePreviewTasks {
       // while
       // resolving lazily through a serializable view — mirrors `composePreviewDiscover` above.
       classpath.from(toolClasspath.incoming.artifactView {}.files)
-      project.configurations.findByName(dependencyConfigName())?.let {
-        classpath.from(it.incoming.artifactView {}.files)
+      if (renderable) {
+        project.configurations.findByName(dependencyConfigName())?.let {
+          classpath.from(it.incoming.artifactView {}.files)
+        }
       }
     }
 

--- a/gradle-plugin/src/main/kotlin/ee/schimke/composeai/plugin/ComposePreviewTasks.kt
+++ b/gradle-plugin/src/main/kotlin/ee/schimke/composeai/plugin/ComposePreviewTasks.kt
@@ -461,14 +461,19 @@ internal object ComposePreviewTasks {
         candidates.firstOrNull { it != null && it.asFile.isDirectory }
       }
 
-    // A non-renderable pure-KMP-Android module has nothing the desktop renderer can pack; skip its
-    // bundle for consistency with the render/daemon tasks (issue #1852).
-    // `isDesktopRenderableConfig`
-    // only trips on the literal `androidRuntimeClasspath` fallback, so this is a no-op on the
-    // Android
-    // bundle path (which routes through `${variant}RuntimeClasspath`) and on real desktop modules.
-    val bundleRenderable = isDesktopRenderableConfig(depConfigName)
     project.tasks.register("composePreviewBundle", BundlePreviewTask::class.java) {
+      // A non-renderable pure-KMP-Android module has nothing the desktop renderer can pack; skip
+      // its
+      // bundle for consistency with the render/daemon tasks (issue #1852). Compute renderability
+      // LAZILY here at task realization — NOT from the eager `depConfigName` snapshot above — so a
+      // cmp-shared module whose `jvm("desktop")` target configures after `registerDesktopTasks`
+      // runs
+      // (when only the androidRuntimeClasspath fallback is visible) isn't permanently skipped
+      // (Codex
+      // review on #1853). Mirrors the render task. `isDesktopRenderableConfig` only trips on the
+      // literal `androidRuntimeClasspath` fallback, so this is a no-op on the Android bundle path
+      // (which routes through `${variant}RuntimeClasspath`) and on real desktop modules.
+      val bundleRenderable = isDesktopRenderableConfig(resolveDependencyConfigName())
       onlyIf { extension.enabled.get() && bundleRenderable }
       previewsJson.set(previewOutputDir.map { it.file("previews.json") })
       moduleClassDirs.from(sourceClassDirs)

--- a/gradle-plugin/src/test/kotlin/ee/schimke/composeai/plugin/KmpAndroidDesktopRoutingTest.kt
+++ b/gradle-plugin/src/test/kotlin/ee/schimke/composeai/plugin/KmpAndroidDesktopRoutingTest.kt
@@ -193,4 +193,35 @@ class KmpAndroidDesktopRoutingTest {
       project.tasks.getByName("validateComposePreviewDesktopRenderClasspath") as TaskInternal
     assertThat(guard.onlyIf.isSatisfiedBy(guard)).isTrue()
   }
+
+  @Test
+  fun `renderability is computed lazily so a desktop target added after registration is not skipped`() {
+    // Codex review on #1853: registerDesktopTasks runs at withPlugin time, BEFORE the consumer's
+    // `kotlin { jvm("desktop") }` block creates desktopRuntimeClasspath. Renderability must be
+    // resolved at task realization, not snapshotted eagerly — otherwise a cmp-shared module whose
+    // desktop config appears later is permanently treated as non-renderable and its render/bundle
+    // are wrongly skipped.
+    val project = ProjectBuilder.builder().withProjectDir(tmp.root).build()
+    val extension = project.extensions.create("composePreview", PreviewExtension::class.java)
+    // Pre-`jvm("desktop")` state: only the androidRuntimeClasspath fallback is visible.
+    project.configurations.create("androidRuntimeClasspath") {
+      isCanBeResolved = true
+      isCanBeConsumed = false
+    }
+
+    ComposePreviewTasks.registerDesktopTasks(project, extension)
+
+    // `jvm("desktop")` configures afterwards, creating the desktop runtime classpath.
+    project.configurations.create("desktopRuntimeClasspath") {
+      isCanBeResolved = true
+      isCanBeConsumed = false
+    }
+
+    // Realized now (after the desktop config exists): render AND bundle must observe it and stay
+    // renderable. A regression to an eager snapshot would skip them.
+    for (taskName in listOf("composePreviewRender", "composePreviewBundle")) {
+      val task = project.tasks.getByName(taskName) as TaskInternal
+      assertThat(task.onlyIf.isSatisfiedBy(task)).isTrue()
+    }
+  }
 }

--- a/gradle-plugin/src/test/kotlin/ee/schimke/composeai/plugin/KmpAndroidDesktopRoutingTest.kt
+++ b/gradle-plugin/src/test/kotlin/ee/schimke/composeai/plugin/KmpAndroidDesktopRoutingTest.kt
@@ -1,6 +1,7 @@
 package ee.schimke.composeai.plugin
 
 import com.google.common.truth.Truth.assertThat
+import org.gradle.api.internal.TaskInternal
 import org.gradle.testfixtures.ProjectBuilder
 import org.junit.Rule
 import org.junit.Test
@@ -129,5 +130,67 @@ class KmpAndroidDesktopRoutingTest {
     discoverTask.dependencyJars.files // resolves; throws on a misnamed config
     assertThat(project.configurations.findByName("desktopRuntimeClasspath")).isNotNull()
     assertThat(project.configurations.findByName("androidRuntimeClasspath")).isNotNull()
+  }
+
+  @Test
+  fun `isDesktopRenderableConfig is false only for the androidRuntimeClasspath fallback`() {
+    // Issue #1852: a pure KMP-Android module whose only runtime config is androidRuntimeClasspath
+    // can't be rendered by the JVM desktop renderer. Every JVM/desktop-flavoured config stays
+    // renderable.
+    assertThat(ComposePreviewTasks.isDesktopRenderableConfig("androidRuntimeClasspath")).isFalse()
+    assertThat(ComposePreviewTasks.isDesktopRenderableConfig("jvmRuntimeClasspath")).isTrue()
+    assertThat(ComposePreviewTasks.isDesktopRenderableConfig("desktopRuntimeClasspath")).isTrue()
+    assertThat(ComposePreviewTasks.isDesktopRenderableConfig("runtimeClasspath")).isTrue()
+  }
+
+  @Test
+  fun `pure KMP-Android module with only androidRuntimeClasspath skips render and guard`() {
+    // Issue #1852: when the desktop fallback lands on androidRuntimeClasspath (no jvm("desktop")
+    // target), the render task and its classpath guard must skip via onlyIf instead of resolving
+    // the ambiguous config and hard-failing the whole pipeline. discover/daemon/bundle skip too.
+    val project = ProjectBuilder.builder().withProjectDir(tmp.root).build()
+    val extension = project.extensions.create("composePreview", PreviewExtension::class.java)
+    project.configurations.create("androidRuntimeClasspath") {
+      isCanBeResolved = true
+      isCanBeConsumed = false
+    }
+
+    ComposePreviewTasks.registerDesktopTasks(project, extension)
+
+    for (taskName in
+      listOf(
+        "composePreviewRender",
+        "validateComposePreviewDesktopRenderClasspath",
+        "composePreviewDiscover",
+        "composePreviewDaemonStart",
+        "composePreviewBundle",
+      )) {
+      val task = project.tasks.getByName(taskName) as TaskInternal
+      assertThat(task.onlyIf.isSatisfiedBy(task)).isFalse()
+    }
+  }
+
+  @Test
+  fun `KMP-Android module with a desktop target stays renderable`() {
+    // The canonical cmp-shared layout (androidRuntimeClasspath + desktopRuntimeClasspath) must NOT
+    // be skipped — desktopRuntimeClasspath wins the fallback and the render task stays enabled.
+    val project = ProjectBuilder.builder().withProjectDir(tmp.root).build()
+    val extension = project.extensions.create("composePreview", PreviewExtension::class.java)
+    project.configurations.create("desktopRuntimeClasspath") {
+      isCanBeResolved = true
+      isCanBeConsumed = false
+    }
+    project.configurations.create("androidRuntimeClasspath") {
+      isCanBeResolved = true
+      isCanBeConsumed = false
+    }
+
+    ComposePreviewTasks.registerDesktopTasks(project, extension)
+
+    val render = project.tasks.getByName("composePreviewRender") as TaskInternal
+    assertThat(render.onlyIf.isSatisfiedBy(render)).isTrue()
+    val guard =
+      project.tasks.getByName("validateComposePreviewDesktopRenderClasspath") as TaskInternal
+    assertThat(guard.onlyIf.isSatisfiedBy(guard)).isTrue()
   }
 }

